tests/contrib/django/test_django_snapshots.py

```diff
@@ -1,9 +1,111 @@
+        c, d = Context(trace_id=1, span_id=1), {}
+        HTTPPropagator.inject(c, d)
+        self._ignore_headers = d
+
+    def _url(self, path):
+        # type: (str) -> str
+        return six.moves.urllib.parse.urljoin(self._base_url, path)
+
+    def get(self, path, **kwargs):
+        return self._session.get(self._url(path), **kwargs)
+
+    def get_ignored(self, path, **kwargs):
+        """Do a normal get request but signal that the trace should be filtered out.
+
+        The signal is a distributed trace id header with the value 1.
+        """
+        headers = kwargs.get("headers", {}).copy()
+        headers.update(self._ignore_headers)
+        kwargs["headers"] = headers
+        return self._session.get(self._url(path), **kwargs)
+
+    def post(self, path, *args, **kwargs):
+        return self._session.post(self._url(path), *args, **kwargs)
+
+    def request(self, method, path, *args, **kwargs):
+        return self._session.request(method, self._url(path), *args, **kwargs)
+
+    def wait(self, path="/", max_tries=100, delay=0.1):
+        # type: (str, int, float) -> None
+        """Wait for the server to start by repeatedly http `get`ting `path` until a 200 is received."""
+
+        @tenacity.retry(stop=tenacity.stop_after_attempt(max_tries), wait=tenacity.wait_fixed(delay))
+        def ping():
+            r = self.get_ignored(path)
+            assert r.status_code == 200
+
+        ping()
+
+
+@pytest.fixture(scope="function")
+def daphne_client(snapshot):
+    """Runs a django app hosted with a daphne webserver in a subprocess and
+    returns a client which can be used to query it.
+
+    Traces are flushed by invoking a tracer.shutdown() using a /shutdown-tracer route
+    at the end of the testcase.
+    """
+
+    # Make sure to copy the environment as we need the PYTHONPATH and _DD_TRACE_WRITER_ADDITIONAL_HEADERS (for the test
+    # token) propagated to the new process.
+    env = os.environ.copy()
+    assert "_DD_TRACE_WRITER_ADDITIONAL_HEADERS" in env, "Client fixture needs test token in headers"
+    env.update(
+        {
+            "DJANGO_SETTINGS_MODULE": "tests.contrib.django.django_app.settings",
+            "DD_TRACE_WRITER_INTERVAL_SECONDS": "0.001",
+        }
+    )
+
+    # ddtrace-run uses execl which replaces the process but the webserver process itself might spawn new processes.
+    # Right now it doesn't but it's possible that it might in the future (ex. uwsgi).
+    cmd = ["ddtrace-run", "daphne", "-p", str(SERVER_PORT), "tests.contrib.django.asgi:application"]
+    proc = subprocess.Popen(
+        cmd,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        close_fds=True,
+        env=env,
+    )
+
+    client = Client("http://localhost:%d" % SERVER_PORT)
+
+    # Wait for the server to start up
+    client.wait()
+
+    try:
+        yield client
+        resp = client.get_ignored("/shutdown-tracer")
```

Comment:
we don't need a try/finally around this because if the context manager block fails then the whole is going to fail anyways?

Reply:
yeah not a bad idea. That'll prevent test cases that run after failing ones from getting messed up :+1:

Reply:
only issue might be if the client is unavailable, then this call might fail too... but... probably bigger issues if that happens.